File: python/promptresponse/beta6.py
"""APR beta.6 JSONC/YAML representation and independent-stream APIs."""

from dataclasses import dataclass
import json
import re
from typing import Any, Iterable, List, Union

import yaml

from .errors import AprParseError
from .models import AprDocument
from .serialization import dumps, loads

VERSION = "1.0-beta.6"


@dataclass(frozen=True)
class Beta6FormRecord:
    """One independent complete form occurrence in a beta.6 stream."""

    document: AprDocument
    # Parsed stream records retain their full semantic JSON object, including
    # extension members that the typed model does not project.
    value: dict[str, Any] | None = None


@dataclass(frozen=True)
class Beta6AttestationRecord:
    """One opaque attestation occurrence carried independently from forms."""

    value: dict[str, Any]


Beta6Record = Union[Beta6FormRecord, Beta6AttestationRecord]


def read_beta6_stream(source: str, representation: str) -> List[Beta6Record]:
    """Read every independent beta.6 occurrence without inferring relationships."""
    if representation == "jsonc":
        raw = [_strip_jsonc(record) for record in _split_jsonc(source)]
    elif representation == "yaml":
        try:
            _reject_yaml_features(source)
            raw = [json.dumps(document, ensure_ascii=False) for document in yaml.load_all(source, AprYamlLoader)]
        except yaml.YAMLError as exc:
            raise AprParseError(f"invalid APR YAML: {exc}") from exc
    else:
        raise ValueError("representation must be 'jsonc' or 'yaml'")
    return [_parse_record(record) for record in raw]


def read_beta6_form(source: str, representation: str) -> AprDocument:
    """Read one form, refusing to silently select a record from a stream."""
    records = read_beta6_stream(source, representation)
    if len(records) != 1 or not isinstance(records[0], Beta6FormRecord):
        raise AprParseError("APR_STREAM_REQUIRES_ITERATION")
    return records[0].document


def write_beta6_form(document: AprDocument, representation: str) -> str:
    """Write a beta.6 form in canonical JSONC or YAML source."""
    if document.version != VERSION:
        raise AprParseError(f"APR beta.6 writers require version {VERSION}")
    return _write_json(dumps(document), representation)


def write_beta6_stream(records: Iterable[Beta6Record], representation: str) -> str:
    """Write every independent stream occurrence in its supplied order."""
    encoded = []
    for record in records:
        json_text = json.dumps(record.value, ensure_ascii=False) if isinstance(record, Beta6FormRecord) and record.value is not None else write_beta6_form(record.document, "jsonc") if isinstance(record, Beta6FormRecord) else json.dumps(record.value, indent=2, ensure_ascii=False)
        encoded.append(_write_json(json_text, representation))
    return "".join("\x1e" + record + "\n" for record in encoded) if representation == "jsonc" else "---\n".join(encoded)


def _parse_record(raw: str) -> Beta6Record:
    try:
        value = json.loads(raw, object_pairs_hook=_unique_object)
    except (json.JSONDecodeError, TypeError) as exc:
        raise AprParseError(f"not valid beta.6 representation: {exc}") from exc
    if not isinstance(value, dict):
        raise AprParseError("an APR beta.6 record must be an object")
    if value.get("version") != VERSION:
        raise AprParseError(f"APR beta.6 records must declare version {VERSION}")
    if "recordType" in value:
        if value["recordType"] != "attestation":
            raise AprParseError("unknown APR beta.6 stream record type")
        _validate_attestation(value)
        return Beta6AttestationRecord(value)
    if "signatures" in value:
        raise AprParseError("RETIRED_EMBEDDED_SIGNATURES")
    return Beta6FormRecord(loads(json.dumps(value, ensure_ascii=False)), value)


def _unique_object(pairs: list[tuple[str, Any]]) -> dict[str, Any]:
    """JSONC object names are unique; never allow parser last-key-wins behavior."""
    value: dict[str, Any] = {}
    for key, item in pairs:
        if key in value:
            raise AprParseError(f"APR JSONC object has duplicate member {key!r}")
        value[key] = item
    return value


def _validate_attestation(value: dict[str, Any]) -> None:
    """Reject records that cannot participate in beta.6 resolution."""
    subject = value.get("subject")
    if not isinstance(subject, dict) or not _digest(subject.get("digest")) or subject.get("canonicalization") != "jcs-sha256":
        raise AprParseError("beta.6 attestation requires subject.digest and jcs-sha256 canonicalization")
    scope = value.get("scope")
    if not isinstance(scope, dict) or scope.get("kind") not in {"document", "fields"}:
        raise AprParseError("beta.6 attestation scope.kind must be document or fields")
    if scope["kind"] == "fields" and (not isinstance(scope.get("fields"), list) or not scope["fields"] or any(not isinstance(field, str) or not field.strip() for field in scope["fields"])):
        raise AprParseError("beta.6 fields attestations require non-blank scope.fields")
    manifest = value.get("manifest")
    if not isinstance(manifest, dict) or not _digest(manifest.get("root")) or not isinstance(manifest.get("entries"), list):
        raise AprParseError("beta.6 attestation requires manifest.root and manifest.entries")
    for entry in manifest["entries"]:
        if not isinstance(entry, dict) or not isinstance(entry.get("path"), str) or not _digest(entry.get("digest")):
            raise AprParseError("beta.6 manifest entries require path and digest")
    if not isinstance(value.get("proofs"), list) or not isinstance(value.get("witnesses"), list) or any(not _digest(witness) for witness in value["witnesses"]):
        raise AprParseError("beta.6 attestations require proofs and digest witnesses arrays")


def _digest(value: Any) -> bool:
    return isinstance(value, str) and len(value) == 71 and value.startswith("sha256:") and all(character in "0123456789abcdef" for character in value[7:])


def _split_jsonc(source: str) -> List[str]:
    records = [record for record in source.split("\x1e") if record.strip()] if "\x1e" in source else [source]
    if not records:
        raise AprParseError("an APR JSONC stream has no records")
    return records


# APR-YAML resolves scalars to the JSON value space, which is not what a stock
# YAML 1.1 loader does. PyYAML resolves "yes" and "on" as booleans, ".inf" as a
# float, "2026-01-01" as a date, and "012" as octal 10 - the last of which is
# silent data corruption in a response. The specification's resolution table is
# implemented here instead: quoted scalars are strings, the null, boolean and
# JSON number forms resolve to those types, and any other plain scalar is a
# string.
_JSON_NUMBER = r"-?(?:0|[1-9][0-9]*)(?:\.[0-9]+)?(?:[eE][-+]?[0-9]+)?"
_NON_FINITE = re.compile(r"^[-+]?\.(?:inf|Inf|INF|nan|NaN|NAN)$")
_DIRECTIVE = re.compile(r"(?m)^%(?:YAML|TAG)\b")


class AprYamlLoader(yaml.SafeLoader):
    """A YAML loader whose scalar resolution is the specification's, not YAML 1.1's."""


AprYamlLoader.yaml_implicit_resolvers = {}
AprYamlLoader.add_implicit_resolver(
    "tag:yaml.org,2002:null", re.compile(r"^(?:~|null|Null|NULL|)$"), ["~", "n", "N", ""]
)
AprYamlLoader.add_implicit_resolver(
    "tag:yaml.org,2002:bool",
    re.compile(r"^(?:true|True|TRUE|false|False|FALSE)$"),
    list("tTfF"),
)
AprYamlLoader.add_implicit_resolver(
    "tag:yaml.org,2002:float",
    re.compile(r"^" + _JSON_NUMBER + r"$"),
    list("-0123456789"),
)


def _reject_non_finite(source: str) -> None:
    """A non-finite float has no JSON value, so it is refused rather than coerced."""
    for line in source.splitlines():
        _, sep, value = line.partition(":")
        if sep and _NON_FINITE.match(value.strip()):
            raise AprParseError(
                "APR YAML forbids a non-finite number: JSON cannot represent it"
            )


def _reject_yaml_features(source: str) -> None:
    if re.search(r"(?m)(?:^|[\s\[{,])(?:[&*!]|<<\s*:)", source):
        raise AprParseError("APR YAML forbids anchors, aliases, tags, and merge keys")
    if _DIRECTIVE.search(source):
        raise AprParseError("APR YAML forbids directives, including %YAML and %TAG")
    _reject_non_finite(source)


def _write_json(source: str, representation: str) -> str:
    if representation == "jsonc":
        return source
    if representation == "yaml":
        return yaml.safe_dump(json.loads(source), allow_unicode=True, sort_keys=False)
    raise ValueError("representation must be 'jsonc' or 'yaml'")


def _strip_jsonc(source: str) -> str:
    result: list[str] = []
    quote = escaped = False
    index = 0
    while index < len(source):
        char = source[index]
        if quote:
            result.append(char)
            if escaped:
                escaped = False
            elif char == "\\":
                escaped = True
            elif char == '"':
                quote = False
            index += 1
            continue
        if char == '"':
            quote = True
            result.append(char)
        elif char == "/" and index + 1 < len(source) and source[index + 1] == "/":
            index = source.find("\n", index)
            if index < 0:
                break
            result.append("\n")
        elif char == "/" and index + 1 < len(source) and source[index + 1] == "*":
            end = source.find("*/", index + 2)
            if end < 0:
                raise AprParseError("unterminated JSONC comment")
            index = end + 1
        else:
            result.append(char)
        index += 1
    import re
    return re.sub(r",(\s*[}\]])", r"\1", "".join(result))

File: typescript/src/beta6.ts
import { parseAllDocuments, stringify } from "yaml";
import { AprParseError } from "./errors.js";
import { AprDocument, JsonObject, JsonValue } from "./model.js";
import { dumps, loads } from "./serialization.js";

/** Source representations defined by APR beta.6. */
export type Beta6Representation = "jsonc" | "yaml";
/** An independent beta.6 stream occurrence. */
export type Beta6Record = { type: "form"; document: AprDocument; /** Full parsed semantic object, including extensions. */ value?: JsonObject } | { type: "attestation"; value: JsonObject };
const VERSION = "1.0-beta.6";

/** Reads all stream occurrences without assigning meaning from their order. */
export function readBeta6Stream(source: string, representation: Beta6Representation): Beta6Record[] {
  const raw = representation === "jsonc" ? splitJsonc(source).map(stripJsonc) : splitYaml(source);
  return raw.map(parseRecord);
}

/** Reads exactly one form; a stream must be handled through readBeta6Stream. */
export function readBeta6Form(source: string, representation: Beta6Representation): AprDocument {
  const records = readBeta6Stream(source, representation);
  if (records.length !== 1 || records[0].type !== "form") throw new AprParseError("APR_STREAM_REQUIRES_ITERATION");
  return records[0].document;
}

/** Writes one beta.6 form in the requested representation. */
export function writeBeta6Form(document: AprDocument, representation: Beta6Representation): string {
  if (document.version !== VERSION) throw new AprParseError(`APR beta.6 writers require version ${VERSION}`);
  const json = dumps(document);
  return representation === "jsonc" ? json : stringify(JSON.parse(json));
}

/** Writes every independent record, preserving occurrence order without linking records. */
export function writeBeta6Stream(records: Iterable<Beta6Record>, representation: Beta6Representation): string {
  const encoded = [...records].map(record => {
    const json = record.type === "form" ? JSON.stringify(record.value ?? JSON.parse(writeBeta6Form(record.document, "jsonc"))) : JSON.stringify(record.value, null, 2);
    return representation === "jsonc" ? json : stringify(JSON.parse(json));
  });
  return representation === "jsonc" ? encoded.map(value => `\u001e${value}\n`).join("") : encoded.join("---\n");
}

function parseRecord(raw: string): Beta6Record {
  let value: unknown;
  try { rejectDuplicateObjectMembers(raw); value = JSON.parse(raw); } catch (error) { throw new AprParseError(`not valid beta.6 representation: ${(error as Error).message}`); }
  if (value === null || typeof value !== "object" || Array.isArray(value)) throw new AprParseError("an APR beta.6 record must be an object");
  const object = value as JsonObject;
  if (object.version !== VERSION) throw new AprParseError(`APR beta.6 records must declare version ${VERSION}`);
  if (object.recordType !== undefined) {
    if (object.recordType !== "attestation") throw new AprParseError("unknown APR beta.6 stream record type");
    validateAttestation(object);
    return { type: "attestation", value: object };
  }
  if (object.signatures !== undefined) throw new AprParseError("RETIRED_EMBEDDED_SIGNATURES");
  return { type: "form", document: loads(JSON.stringify(object)), value: object };
}

/** Lightweight JSON member scanner used before JSON.parse's otherwise silent last-key-wins behavior. */
function rejectDuplicateObjectMembers(source: string): void {
  const stack: Array<Set<string> | null> = [];
  let i = 0;
  const whitespace = /\s/;
  const string = (): string => {
    const start = i++; let escaped = false;
    while (i < source.length) { const c = source[i++]; if (escaped) escaped = false; else if (c === "\\") escaped = true; else if (c === '"') return JSON.parse(source.slice(start, i)); }
    throw new Error("unterminated JSON string");
  };
  while (i < source.length) {
    const c = source[i]; if (whitespace.test(c) || c === "," || c === ":") { i++; continue; }
    if (c === "{") { stack.push(new Set()); i++; continue; }
    if (c === "[") { stack.push(null); i++; continue; }
    if (c === "}" || c === "]") { stack.pop(); i++; continue; }
    if (c === '"') {
      const value = string(); let next = i; while (next < source.length && whitespace.test(source[next])) next++;
      const object = stack.at(-1);
      if (source[next] === ":" && object) { if (object.has(value)) throw new Error(`duplicate member ${value}`); object.add(value); }
      continue;
    }
    i++;
  }
}

function validateAttestation(value: JsonObject): void {
  const subject = isObject(value.subject) ? value.subject : undefined;
  if (!subject || !isDigest(subject.digest) || subject.canonicalization !== "jcs-sha256") throw new AprParseError("beta.6 attestation requires subject.digest and jcs-sha256 canonicalization");
  const scope = isObject(value.scope) ? value.scope : undefined;
  if (!scope || (scope.kind !== "document" && scope.kind !== "fields")) throw new AprParseError("beta.6 attestation scope.kind must be document or fields");
  if (scope.kind === "fields" && (!Array.isArray(scope.fields) || !scope.fields.length || scope.fields.some(field => typeof field !== "string" || !field.trim()))) throw new AprParseError("beta.6 fields attestations require non-blank scope.fields");
  const manifest = isObject(value.manifest) ? value.manifest : undefined;
  if (!manifest || !isDigest(manifest.root) || !Array.isArray(manifest.entries)) throw new AprParseError("beta.6 attestation requires manifest.root and manifest.entries");
  for (const entry of manifest.entries) if (!isObject(entry) || typeof entry.path !== "string" || !isDigest(entry.digest)) throw new AprParseError("beta.6 manifest entries require path and digest");
  if (!Array.isArray(value.proofs) || !Array.isArray(value.witnesses) || value.witnesses.some(witness => !isDigest(witness))) throw new AprParseError("beta.6 attestations require proofs and digest witnesses arrays");
}

function isDigest(value: JsonValue | undefined): value is string {
  return typeof value === "string" && /^sha256:[0-9a-f]{64}$/.test(value);
}

function isObject(value: JsonValue | undefined): value is JsonObject {
  return value !== null && typeof value === "object" && !Array.isArray(value);
}

function splitJsonc(source: string): string[] {
  const records = source.includes("\u001e") ? source.split("\u001e").filter(Boolean) : [source];
  if (!records.length) throw new AprParseError("an APR JSONC stream has no records");
  return records;
}

function splitYaml(source: string): string[] {
  if (/(^|[\s\[{,])(?:[&*!]|<<\s*:)/m.test(source)) throw new AprParseError("APR YAML forbids anchors, aliases, tags, and merge keys");
  const documents = parseAllDocuments(source, { schema: "core" });
  return documents.map(document => {
    if (document.errors.length) throw new AprParseError(`invalid APR YAML: ${document.errors[0].message}`);
    return JSON.stringify(document.toJSON() as JsonValue);
  });
}

function stripJsonc(source: string): string {
  let result = "", quote = false, escaped = false;
  for (let i = 0; i < source.length; i++) {
    const c = source[i];
    if (quote) { result += c; if (escaped) escaped = false; else if (c === "\\") escaped = true; else if (c === '"') quote = false; continue; }
    if (c === '"') { quote = true; result += c; continue; }
    if (c === "/" && source[i + 1] === "/") { while (i < source.length && source[i] !== "\n") i++; result += "\n"; continue; }
    if (c === "/" && source[i + 1] === "*") { i += 2; while (i + 1 < source.length && !(source[i] === "*" && source[i + 1] === "/")) i++; if (i + 1 >= source.length) throw new AprParseError("unterminated JSONC comment"); i++; continue; }
    result += c;
  }
  return result.replace(/,(\s*[}\]])/g, "$1");
}
